src/shared/Services/FileCacheService.cs

```diff
@@ -0,0 +1,117 @@
+        if (fileValue is not null)
+        {
+            return fileValue;
+        }
+
+        // it's not cached so create it and add it to the file cache if it's not null
+        _logger.LogWarning("File cache miss for {key}", key.SanitizeForLog());
+        var newValue = await factory();
+
+        if (newValue is not null)
+        {
+            await SetValueAsync(key, newValue, token);
+        }
+
+        return newValue;
+    }
+
+    public async Task<TItem?> GetValueAsync<TItem>(string key, CancellationToken token)
+    {
+        var filePath = GetFilePath(key).SanitizePath(_cacheDirectory);
+        if (File.Exists(filePath))
+        {
+            var item = await File.ReadAllTextAsync(filePath, token);
+
+            if (typeof(TItem) == typeof(string))
+            {
+                return (TItem)(object)item; //coerce the string to TItem
+            }
+
+            // otherwise assume it's json and deserialize it
+            return item.ToObject<TItem>();
+        }
+
+        return default;
+    }
+
+    public async Task SetValueAsync<TItem>(string key, TItem? value, CancellationToken token)
+    {
+        if (value is not null)
+        {
+            var filePath = GetFilePath(key).SanitizePath(_cacheDirectory);
+            if (typeof(TItem) == typeof(string))
+            {
+                await File.WriteAllTextAsync(filePath, value.ToString(), token);
+            }
+            else
+            {
+                await File.WriteAllTextAsync(filePath, value.ToJson(), token);
+            }
+
+            _logger.LogWarning("Cached {Key} of type {Type}", key, typeof(TItem).Name);
+        }
+    }
+
+    public void Clear()
+    {
+        if (Directory.Exists(_cacheDirectory))
+        {
+            foreach (var file in Directory.GetFiles(_cacheDirectory, "*"))
+            {
+                var sanitizedPath = file.SanitizePath(_cacheDirectory);
+                _logger.LogInformation("Deleting file {File}", sanitizedPath);
+                File.Delete(sanitizedPath);
+            }
+        }
+    }
+
+    public void RemoveStore(string storeId)
+    {
+        _logger.LogInformation("Invalidating store {storeId}", storeId);
+        if (Directory.Exists(_cacheDirectory))
+        {
+            foreach (var file in Directory.GetFiles(_cacheDirectory, $"{storeId}*"))
+            {
+                var sanitizedPath = file.SanitizePath(_cacheDirectory);
+                _logger.LogInformation("Deleting file {File}", sanitizedPath);
+                File.Delete(sanitizedPath);
+            }
+        }
+    }
+
+    public void RemoveKey(string key)
+    {
+        var filePath = GetFilePath(key).SanitizePath(_cacheDirectory);
+        if (File.Exists(filePath))
+        {
+            _logger.LogInformation("Deleting file {File}", filePath);
+            File.Delete(filePath);
+        }
+    }
+
+    private string GetFilePath(string key) => Path.Combine(_cacheDirectory, key);
```

Comment:
this key would need to be the cachekey hash

Reply:
THinking through it since just a few mins ago, i think we should do it like this

pass in the cached-key, which is "storeId-roothash-keyhex" then md5 the entire cache key, instead of just the key itself.

Then filecache will just do the hash transform when it writes the file or reads the file, that way the rest of the app doesnt have to care about hashing cache keys and we are sure to get predicatable filenames that are not to long

Reply:
I i think the folder strucutre will be storeId only as we need an easy way to purge a store's cache and we might have lost the root hash it started with.

Reply:
including storeId in the key is redundant since the file will be in a storeId folder.